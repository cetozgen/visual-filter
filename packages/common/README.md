# visual-filter/common

Shared utilities for the @visual-filter packages
