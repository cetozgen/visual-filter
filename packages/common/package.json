{
  "name": "@visual-filter/common",
  "version": "1.0.3",
  "description": "Shared utilities for the @visual-filter packages",
  "keywords": [
    "visual-filter"
  ],
  "homepage": "https://github.com/obadakhalili/vue-visual-filter.git",
  "repository": {
    "type": "git",
    "url": "https://github.com/obadakhalili/vue-visual-filter.git"
  },
  "license": "MIT",
  "author": "Obada Khalili",
  "main": "dist/index.js",
  "types": "dist/index.d.ts",
  "files": [
    "dist"
  ],
  "scripts": {
    "build": "tsc --project tsconfig.build.json",
    "clean": "rm -rf dist",
    "dev": "yarn build --watch",
    "format": "yarn run prettier --ignore-path ../../.gitignore --write",
    "preversion": "yarn build"
  },
  "devDependencies": {
    "prettier": "2.7.1",
    "typescript": "4.7.4"
  }
}
