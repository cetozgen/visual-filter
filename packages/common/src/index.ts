export enum FilterType {
  GROUP = "group",
  CONDITION = "condition",
}

export enum GroupType {
  AND = "and",
  NOT_AND = "not and",
  OR = "or",
  NOT_OR = "not or",
}

export enum DataType {
  NUMERIC = "numeric",
  NOMINAL = "nominal",
}

export function deepCopy(src: any): any {
  if (Object(src) !== src) {
    return src
  }
  if (src.constructor === Object) {
    return Object.entries(src).reduce(
      (objCopy, [key, value]) => ({ ...objCopy, [key]: deepCopy(value) }),
      {},
    )
  }
  return src.map(deepCopy)
}
