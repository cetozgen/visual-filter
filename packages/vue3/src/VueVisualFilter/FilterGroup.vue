<script>
export default {
  name: "FilterGroup",
  emits: ["addFilter", "deleteGroup"],
  props: {
    group: {
      type: Object,
      required: true,
      validator(value) {
        return value.constructor === Object
      },
    },
    filterTypes: {
      type: Array,
      required: true,
    },
    groupTypes: {
      type: Array,
      required: true,
    },
    removable: {
      type: Boolean,
      required: true,
    },
  },
  methods: {
    addFilter(newFilterType) {
      if (this.filterTypes.includes(newFilterType)) {
        this.$emit("addFilter", this.group.filters, newFilterType)
      }
    },
  },
}
</script>

<template>
  <div class="space-y-4">
    <div class="space-x-2">
      <slot name="groupTypes" v-bind="{ groupTypes, group }">
        <select v-model="group.groupType">
          <option v-for="type in groupTypes" :key="type" :value="type">
            {{ type }}
          </option>
        </select>
      </slot>
      <slot name="filterAddition" v-bind="{ filterTypes, addFilter }">
        <select @change="addFilter($event.target.value)">
          <option v-for="type in filterTypes" :key="type" :value="type">
            {{ type }}
          </option>
        </select>
      </slot>
      <slot
        v-if="removable"
        name="groupDeletion"
        :deleteGroup="() => $emit('deleteGroup', group)"
      >
        <button @click="$emit('deleteGroup', group)">x</button>
      </slot>
    </div>
    <div v-if="group.filters.length" class="ml-10 space-y-1">
      <slot name="groupChildren"></slot>
    </div>
  </div>
</template>
