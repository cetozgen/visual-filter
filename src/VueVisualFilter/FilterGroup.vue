<script>
export default {
  name: "FilterGroup",
  emits: ["updateType", "addFilter", "deleteGroup"],
  props: {
    group: {
      type: Object,
      required: true,
      validator(value) {
        return value.constructor === Object
      }
    },
    filterTypes: {
      type: Array,
      required: true
    },
    groupTypes: {
      type: Array,
      required: true
    },
    removable: {
      type: Boolean,
      required: true
    }
  },
  methods: {
    updateType(newGroupType) {
      if (this.groupTypes.includes(newGroupType)) {
        this.$emit("updateType", this.group, newGroupType)
      }
    },
    addFilter(newFilterType) {
      if (this.filterTypes.includes(newFilterType)) {
        this.$emit("addFilter", this.group.filters, newFilterType)
      }
    }
  }
}
</script>

<template>
  <div class="space-y-4">
    <div class="space-x-2">
      <slot name="groupTypes" :groupTypes="groupTypes" :updateType="updateType">
        <select @change="updateType($event.target.value)">
          <option v-for="type in groupTypes" :key="type" :value="type">
            {{ type }}
          </option>
        </select>
      </slot>
      <slot
        name="filterAddition"
        :filterTypes="filterTypes"
        :addFilter="addFilter"
      >
        <select @change="addFilter($event.target.value)">
          <option v-for="type in filterTypes" :key="type" :value="type">
            {{ type }}
          </option>
        </select>
      </slot>
      <slot
        v-if="removable"
        name="filterDeletion"
        :deleteGroup="() => $emit('deleteGroup', group)"
      >
        <button @click="$emit('deleteGroup', group)">x</button>
      </slot>
    </div>
    <div v-if="group.filters.length" class="ml-10 space-y-1">
      <slot name="groupChildren"></slot>
    </div>
  </div>
</template>
