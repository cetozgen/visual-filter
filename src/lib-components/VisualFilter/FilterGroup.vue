<script>
export default {
  name: "FilterGroup",
  emits: ["addFilter", "changeGroupType"],
  props: ["group"],
  data() {
    return {
      groupTypes: ["and", "not and", "or", "not or"]
    }
  },
  methods: {
    addFilter(e) {
      const newFilterType = e.target.value

      if (newFilterType === "group" || newFilterType === "condition") {
        this.$emit("addFilter", this.group.filters, newFilterType)
      }
    },
    changeGroupType(e) {
      const newGroupType = e.target.value

      if (this.groupTypes.includes(newGroupType)) {
        this.$emit("changeGroupType", this.group, newGroupType)
      }
    }
  }
}
</script>

<template>
  <select @change="changeGroupType" class="mr-2">
    <option v-for="type in this.groupTypes" :key="type" :value="type">
      {{ type }}
    </option>
  </select>
  <select @change="addFilter">
    <option value="group">Group</option>
    <option value="condition">Condition</option>
  </select>
  <div v-if="group.filters.length" class="ml-5 mt-2">
    <slot></slot>
  </div>
</template>
