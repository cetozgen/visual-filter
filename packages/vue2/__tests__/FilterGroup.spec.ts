import { mount } from "@vue/test-utils"
import { FilterType, GroupType } from "@visual-filter/common"

import FilterGroup from "../src/VueVisualFilter/FilterGroup.vue"

const filterTypes = Object.values(FilterType)
const groupTypes = Object.values(GroupType)

const propsData = {
  group: {
    filters: [],
    groupType: GroupType.AND,
    type: FilterType.GROUP,
  },
  removable: false,
}
const wrapper = mount<FilterGroup & { [name: string]: any }>(FilterGroup, {
  propsData,
})

describe("group types logic", () => {
  it("should update modeled property", async () => {
    const selectedGroupIndex = 1
    await wrapper
      .find("select")
      .findAll("option")
      .at(selectedGroupIndex)
      .setSelected()
    expect(propsData.group.groupType).toBe(groupTypes[selectedGroupIndex])
  })
})

describe("filter addition logic", () => {
  const selectedFilterIndex = 1
  const addFilterSpy = jest.spyOn(wrapper.vm, "addFilter")

  beforeAll(async () => {
    await wrapper
      .findAll("select")
      .at(1)
      .findAll("option")
      .at(selectedFilterIndex)
      .setSelected()
  })

  it("should call addFilter method on change event", () => {
    expect(addFilterSpy).toHaveBeenCalled()
  })

  it("should emit addFilter on change event with correct paramaters", () => {
    const emittedEvents = wrapper.emitted()
    const [params] = emittedEvents.addFilter as unknown[][]
    expect(params).toEqual([
      propsData.group.filters,
      filterTypes[selectedFilterIndex],
    ])
  })
})
