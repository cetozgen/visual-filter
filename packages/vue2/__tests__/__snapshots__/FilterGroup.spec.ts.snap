// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`filter addition logic should emit addFilter on change event with correct paramaters 1`] = `
Array [
  Array [],
  "condition",
]
`;

exports[`group types logic should update modeled property 1`] = `"not and"`;

exports[`slots should bound correct values to their corresponding slots 1`] = `
Array [
  "group",
  "condition",
]
`;
